crates/wasmparser/src/validator/operators.rs

```diff
@@ -310,1885 +380,3521 @@ impl OperatorValidator {
-            return Err(OperatorValidatorError::new(
-                "bulk memory support is not enabled",
-            ));
+            bail_op_err!(offset, "bulk memory support is not enabled");
         }
         Ok(())
     }
 
-    fn check_shared_memarg_wo_align(
+    fn check_shared_memarg_wo_align<T>(
         &self,
+        offset: usize,
+        resources: &T,
         memarg: MemoryImmediate,
-        resources: impl WasmModuleResources,
-    ) -> OperatorValidatorResult<ValType> {
-        self.check_memory_index(memarg.memory, resources)
+    ) -> OperatorValidatorResult<ValType>
+    where
+        T: WasmModuleResources,
+    {
+        self.check_memory_index(offset, resources, memarg.memory)
     }
 
-    fn check_simd_lane_index(&self, index: SIMDLaneIndex, max: u8) -> OperatorValidatorResult<()> {
+    fn check_simd_lane_index(
+        &self,
+        offset: usize,
+        index: SIMDLaneIndex,
+        max: u8,
+    ) -> OperatorValidatorResult<()> {
         if index >= max {
-            return Err(OperatorValidatorError::new("SIMD index out of bounds"));
+            bail_op_err!(offset, "SIMD index out of bounds");
         }
         Ok(())
     }
 
     /// Validates a block type, primarily with various in-flight proposals.
-    fn check_block_type(
+    fn check_block_type<T>(
         &self,
+        offset: usize,
+        resources: &T,
         ty: BlockType,
-        resources: impl WasmModuleResources,
-    ) -> OperatorValidatorResult<()> {
+    ) -> OperatorValidatorResult<()>
+    where
+        T: WasmModuleResources,
+    {
         match ty {
             BlockType::Empty
             | BlockType::Type(ValType::I32)
             | BlockType::Type(ValType::I64)
             | BlockType::Type(ValType::F32)
             | BlockType::Type(ValType::F64) => Ok(()),
             BlockType::Type(ValType::ExternRef) | BlockType::Type(ValType::FuncRef) => {
-                self.check_reference_types_enabled()
+                self.check_reference_types_enabled(offset)
             }
-            BlockType::Type(ValType::V128) => self.check_simd_enabled(),
+            BlockType::Type(ValType::V128) => self.check_simd_enabled(offset),
             BlockType::FuncType(idx) => {
                 if !self.features.multi_value {
-                    return Err(OperatorValidatorError::new(
+                    bail_op_err!(
+                        offset,
                         "blocks, loops, and ifs may only produce a resulttype \
                          when multi-value is not enabled",
-                    ));
+                    );
                 }
-                func_type_at(&resources, idx)?;
+                func_type_at(resources, idx, offset)?;
                 Ok(())
             }
         }
     }
 
     /// Validates a `call` instruction, ensuring that the function index is
     /// in-bounds and the right types are on the stack to call the function.
-    fn check_call(
+    fn check_call<T>(
```

Comment:
How come the usage of `impl WasmModuleResources` changed to `<T>` on many methods?

Reply:
Haha, interesting catch! This is an artifact of the refactorings that went through a phase where `OperatorValidator` had ownership of `resources` and then lost it again due to aforementioned problems and then due to my own personal preferences for the `<T>` syntax this syntax evolved. We can change it back if you dislike the new syntax.

Reply:
I reverted these syntactical changes.